chainstate/tx-verifier/src/transaction_verifier/input_check/mod.rs

```diff
@@ -0,0 +1,302 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_types::block_index_ancestor_getter;
+use common::{
+    chain::{
+        block::timestamp::BlockTimestamp,
+        signature::{inputsig::InputWitness, Signable, Transactable},
+        ChainConfig, GenBlock, TxInput, TxOutput,
+    },
+    primitives::{BlockHeight, Id},
+};
+use mintscript::{InputInfo, SignatureContext, TimelockContext};
+
+use crate::TransactionVerifierStorageRef;
+
+use super::{error::ConnectTransactionError, TransactionSourceForConnect};
+
+// TODO(PR): Maybe fuse block and transaction contexts into one
+pub struct BlockVerificationContext<'a, S, AW> {
+    chain_config: &'a ChainConfig,
+    spending_time: BlockTimestamp,
+    spending_height: BlockHeight,
+    tip: Id<GenBlock>,
+    storage: &'a S,
+    pos_accounting: &'a AW,
+}
+
+impl<'a, S, AW> BlockVerificationContext<'a, S, AW> {
+    pub fn from_source(
+        chain_config: &'a ChainConfig,
+        spending_time: BlockTimestamp,
+        tx_source: &TransactionSourceForConnect,
+        storage: &'a S,
+        pos_accounting: &'a AW,
+    ) -> Self {
+        let tip = match tx_source {
+            TransactionSourceForConnect::Chain { new_block_index } => {
+                (*new_block_index.block_id()).into()
+            }
+            TransactionSourceForConnect::Mempool {
+                current_best,
+                effective_height: _,
+            } => current_best.block_id(),
+        };
+
+        Self::custom(
+            chain_config,
+            spending_time,
+            tx_source.expected_block_height(),
+            tip,
+            storage,
+            pos_accounting,
+        )
+    }
+
+    pub fn custom(
+        chain_config: &'a ChainConfig,
+        spending_time: BlockTimestamp,
+        spending_height: BlockHeight,
+        tip: Id<GenBlock>,
+        storage: &'a S,
+        pos_accounting: &'a AW,
+    ) -> Self {
+        Self {
+            chain_config,
+            spending_time,
+            spending_height,
+            tip,
+            storage,
+            pos_accounting,
+        }
+    }
+}
+
+pub struct TransactionVerificationContext<'a, T, S, AW> {
+    block_ctx: &'a BlockVerificationContext<'a, S, AW>,
+    transactable: &'a T,
+    inputs_and_sigs: Vec<(InputInfo<'a>, &'a InputWitness)>,
+}
+
+impl<'a, T: Signable + Transactable, S, AW> TransactionVerificationContext<'a, T, S, AW> {
+    pub fn new<UW: utxo::UtxosView>(
+        block_ctx: &'a BlockVerificationContext<'a, S, AW>,
+        utxo_view: &UW,
+        transactable: &'a T,
+    ) -> Result<Self, ConnectTransactionError> {
+        let inputs = transactable.inputs().unwrap_or_default();
+        let sigs = transactable.signatures().unwrap_or_default();
+
+        // TODO(PR): Should this be a proper check rather than an assertion? Is it correct?
+        assert_eq!(inputs.len(), sigs.len());
```

Comment:
I don't mind keeping this here. There's a check upon decoding a SignedTransaction that these are equal.